File: Game/IRTaktiks/IRTaktiks/Listener/TuioClient.cs
using System;
using System.Threading;
using System.Collections;
using IRTaktiks.Listener.Oscpack;

namespace IRTaktiks.Listener
{
    public struct tuio
    {
        public int f_id;
        public float xpos, ypos, angle;

        public tuio(int id, float x, float y, float a)
        {
            this.f_id = id;
            this.xpos = x;
            this.ypos = y;
            this.angle = a;
        }

        public void update(float x, float y, float a)
        {
            this.xpos = x;
            this.ypos = y;
            this.angle = a;
        }
    }

    public class TuioClient
    {
        private bool listening = false;
        private int port = 3333;
        private OSCReceiver receiver;
        private Thread thread;

        private Hashtable objectList = new Hashtable(32);
        private ArrayList aliveObjectList = new ArrayList(32);
        private ArrayList newObjectList = new ArrayList(32);
        private ArrayList aliveCursorList = new ArrayList(16);
        private ArrayList newCursorList = new ArrayList(16);

        private int currentFrame = 0;
        private int lastFrame = 0;

        private ArrayList listenerList = new ArrayList();

        public TuioClient() { }

        public TuioClient(int port)
        {
            this.port = port;
        }

        public int getPort()
        {
            return port;
        }

        public void connect()
        {
            try
            {
                receiver = new OSCReceiver(port);
                startListening();
            }
            catch
            { }
        }

        public void disconnect()
        {
            stopListening();
        }

        private void startListening()
        {
            listening = true;
            thread = new Thread(new ThreadStart(listen));
            thread.Start();
        }

        private void stopListening()
        {
            receiver.Close();
            listening = false;
            receiver = null;
        }

        private void listen()
        {
            while (listening)
            {
                try
                {
                    OSCPacket packet = receiver.Receive();
                    if (packet != null)
                    {
                        if (packet.IsBundle())
                        {
                            ArrayList messages = packet.Values;
                            for (int i = 0; i < messages.Count; i++)
                            {
                                processMessage((OSCMessage)messages[i]);
                            }
                        }
                        else processMessage((OSCMessage)packet);
                    }
                    else Console.WriteLine("null packet");
                }
                catch (Exception)
                { }
            }
        }

        private void processMessage(OSCMessage message)
        {
            string address = message.Address;
            ArrayList args = message.Values;
            string command = (string)args[0];

            if (address == "/tuio/2Dobj")
            {
                if ((command == "set") && (currentFrame >= lastFrame))
                {
                    int s_id = (int)args[1];
                    int f_id = (int)args[2];
                    float x = (float)args[3];
                    float y = (float)args[4];
                    float a = (float)args[5];
                    float X = (float)args[6];
                    float Y = (float)args[7];
                    float A = (float)args[8];
                    float m = (float)args[9];
                    float r = (float)args[10];

                    //Console.WriteLine(s_id+" "+f_id+" "+x+" "+y+" "+a+" "+X+" "+Y+" "+A+" "+m+" "+r);

                    if (!objectList.ContainsKey(args[1]))
                    {
                        tuio t = new tuio(f_id, x, y, a);
                        objectList.Add(s_id, t);

                        for (int i = 0; i < listenerList.Count; i++)
                        {
                            TuioListener listener = (TuioListener)listenerList[i];
                            
                            if (listener != null)
                            {
                                listener.addTuioObj(s_id, f_id, x, y, a, X, Y, A, m, r);
                                listener.updateTuioObj(s_id, f_id, x, y, a, X, Y, A, m, r);
                            }
                        }
                    }
                    else
                    {
                        tuio t = (tuio)objectList[s_id];
                        if ((t.xpos != x) || (t.ypos != y) || (t.angle != a))
                        {
                            for (int i = 0; i < listenerList.Count; i++)
                            {
                                TuioListener listener = (TuioListener)listenerList[i];
                                if (listener != null) listener.updateTuioObj(s_id, f_id, x, y, a, X, Y, A, m, r);
                            }
                            //Console.WriteLine(t.xpos +" " + t.ypos + " " + t.angle);
                            t.update(x, y, a);
                            objectList[s_id] = t;
                        }
                    }

                }
                else if ((command == "alive") && (currentFrame >= lastFrame))
                {
                    for (int i = 1; i < args.Count; i++)
                    {
                        // get the message content
                        newObjectList.Add((int)args[i]);
                        // reduce the object list to the lost objects
                        if (aliveObjectList.Contains(args[i]))
                            aliveObjectList.Remove(args[i]);
                    }

                    // remove the remaining objects
                    for (int i = 0; i < aliveObjectList.Count; i++)
                    {
                        int s_id = (int)aliveObjectList[i];
                        int f_id = ((tuio)objectList[aliveObjectList[i]]).f_id;
                        objectList.Remove(aliveObjectList[i]);
                        //Console.WriteLine("remove "+f_id);
                        for (int j = 0; j < listenerList.Count; j++)
                        {
                            TuioListener listener = (TuioListener)listenerList[j];
                            if (listener != null) listener.removeTuioObj(s_id, f_id);
                        }
                    }

                    ArrayList buffer = aliveObjectList;
                    aliveObjectList = newObjectList;

                    // recycling of the ArrayList
                    newObjectList = buffer;
                    newObjectList.Clear();

                }
                else if (command == "fseq")
                {
                    lastFrame = currentFrame;
                    currentFrame = (int)args[1];
                    if (currentFrame == -1) currentFrame = lastFrame;

                    if (currentFrame >= lastFrame)
                    {
                        for (int i = 0; i < listenerList.Count; i++)
                        {
                            TuioListener listener = (TuioListener)listenerList[i];
                            if (listener != null) listener.refresh();
                        }
                    }
                }

            }

            else if (address == "/tuio/2Dcur")
            {

                if ((command == "set") && (currentFrame >= lastFrame))
                {
                    int s_id = (int)args[1];
                    float x = (float)args[2];
                    float y = (float)args[3];
                    float X = (float)args[4];
                    float Y = (float)args[5];
                    float m = (float)args[6];

                    for (int i = 0; i < listenerList.Count; i++)
                    {
                        TuioListener listener = (TuioListener)listenerList[i];
                        if (listener != null) listener.updateTuioCur(s_id, x, y, X, Y, m);
                    }
                }
                else if ((command == "alive") && (currentFrame >= lastFrame))
                {

                    for (int i = 1; i < args.Count; i++)
                    {
                        // get the message content
                        newCursorList.Add(args[i]);
                        
                        // reduce the object list to the lost objects
                        if (aliveCursorList.Contains(args[i]))
                            aliveCursorList.Remove(args[i]);
                        else
                        {
                            for (int j = 0; j < listenerList.Count; j++)
                            {
                                TuioListener listener = (TuioListener)listenerList[j];
                                if (listener != null)
                                    listener.addTuioCur((int)args[i]);
                            }
                        }
                    }

                    // remove the remaining objects
                    for (int i = 0; i < aliveCursorList.Count; i++)
                    {
                        int s_id = (int)aliveCursorList[i];
                        //Console.WriteLine("remove "+s_id);
                        for (int j = 0; j < listenerList.Count; j++)
                        {
                            TuioListener listener = (TuioListener)listenerList[j];
                            if (listener != null) listener.removeTuioCur(s_id);
                        }
                    }

                    ArrayList buffer = aliveCursorList;
                    aliveCursorList = newCursorList;

                    // recycling of the ArrayList
                    newCursorList = buffer;
                    newCursorList.Clear();
                }
            }
        }

        public void addTuioListener(TuioListener listener)
        {
            listenerList.Add(listener);
        }

        public void removeTuioListener(TuioListener listener)
        {
            listenerList.Remove(listener);
        }
    }
}

File: Code/Reactivision/TuioClient/TuioClient.cpp
/*
	TUIO C++ backend - part of the reacTIVision project
	http://mtg.upf.es/reactable

	Copyright (c) 2006 Martin Kaltenbrunner <mkalten@iua.upf.es>

	Permission is hereby granted, free of charge, to any person obtaining
	a copy of this software and associated documentation files
	(the "Software"), to deal in the Software without restriction,
	including without limitation the rights to use, copy, modify, merge,
	publish, distribute, sublicense, and/or sell copies of the Software,
	and to permit persons to whom the Software is furnished to do so,
	subject to the following conditions:

	The above copyright notice and this permission notice shall be
	included in all copies or substantial portions of the Software.

	Any person wishing to distribute modifications to the Software is
	requested to send the modifications to the original developer so that
	they can be incorporated into the canonical version.

	THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
	EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
	MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
	IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR
	ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF
	CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
	WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/

#include "TuioClient.h"


#ifndef WIN32
static void* ThreadFunc( void* obj )
#else
static DWORD WINAPI ThreadFunc( LPVOID obj )
#endif
{
	static_cast<TuioClient*>(obj)->socket->Run();
	return 0;
};
	
TuioClient::TuioClient() {
	TuioClient(3333);
}

TuioClient::TuioClient(int port) {
	socket = new UdpListeningReceiveSocket(IpEndpointName( IpEndpointName::ANY_ADDRESS, port ), this );
	locked = false;
}

TuioClient::~TuioClient() {
	delete socket;
}

void TuioClient::ProcessBundle( const ReceivedBundle& b, const IpEndpointName& remoteEndpoint) {
	for( ReceivedBundle::const_iterator i = b.ElementsBegin(); i != b.ElementsEnd(); ++i ){
		if( i->IsBundle() )
			ProcessBundle( ReceivedBundle(*i), remoteEndpoint);
		else
			ProcessMessage( ReceivedMessage(*i), remoteEndpoint);
	}
}


void TuioClient::ProcessMessage( const ReceivedMessage& msg, const IpEndpointName& remoteEndpoint) {
	try {
		ReceivedMessageArgumentStream args = msg.ArgumentStream();
		ReceivedMessage::const_iterator arg = msg.ArgumentsBegin();

		if( strcmp( msg.AddressPattern(), "/tuio/2Dobj" ) == 0 ){

			const char* cmd;
			args >> cmd;
			
			if( strcmp( cmd, "set" ) == 0 ){	
				int32 s_id, f_id;
				float xpos, ypos, angle, xspeed, yspeed, rspeed, maccel, raccel;
					
				args >> s_id >> f_id >> xpos >> ypos >> angle >> xspeed >> yspeed >> rspeed >> maccel >> raccel >> EndMessage;

				std::list<tuio>::iterator iter;
				for (iter=objectList.begin(); iter != objectList.end(); iter++)
					if(iter->s_id==(unsigned int)s_id) break;

				if (iter == objectList.end()) {
					listener->addTuioObj((unsigned long)s_id,(unsigned int)f_id);
					listener->updateTuioObj((unsigned int)s_id,(unsigned int)f_id,xpos,ypos,angle,xspeed,yspeed,rspeed,maccel,raccel);

					tuio add_tuio;
					add_tuio.s_id=(unsigned int)s_id;
					add_tuio.f_id=(unsigned int)f_id;
					add_tuio.xpos = xpos;
					add_tuio.ypos = ypos;
					add_tuio.angle = angle;
					objectList.push_back(add_tuio);
				} else if ( (iter->xpos!=xpos) || (iter->ypos!=ypos) || (iter->angle!=angle) ) {
					listener->updateTuioObj((unsigned int)s_id,(unsigned int)f_id,xpos,ypos,angle,xspeed,yspeed,rspeed,maccel,raccel);
					iter->xpos = xpos;
					iter->ypos = ypos;
					iter->angle = angle;
				}

			} else if( strcmp( cmd, "alive" ) == 0 ){

				int32 s_id;
				while(!args.Eos()) {
					args >> s_id;
					objectBuffer.push_back((unsigned int)s_id);
					
					std::list<unsigned int>::iterator iter;
					iter = find(aliveObjectList.begin(), aliveObjectList.end(), (unsigned int)s_id); 
					if (iter != aliveObjectList.end()) aliveObjectList.erase(iter);
				}
				args >> EndMessage;
				
				std::list<unsigned int>::iterator alive_iter;
				for (alive_iter=aliveObjectList.begin(); alive_iter != aliveObjectList.end(); alive_iter++) {
					std::list<tuio>::iterator object_iter;
					for (object_iter=objectList.begin(); object_iter != objectList.end(); object_iter++) {
						if(object_iter->s_id==*alive_iter) {
							listener->removeTuioObj(object_iter->s_id,object_iter->f_id);
							objectList.erase(object_iter);
							break;
						}
					}
					
				}
			
				//std::list<unsigned int> newObjectList = aliveObjectList;
				aliveObjectList = objectBuffer;
			
				// recycling the list
				//objectBuffer = newObjectList;
				objectBuffer.clear();
			} else if( strcmp( cmd, "fseq" ) == 0 ){
				
				int32 lastFrame = currentFrame;
				args >> currentFrame  >> EndMessage;
				if (currentFrame>lastFrame) listener->refresh();
			}
		} else if( strcmp( msg.AddressPattern(), "/tuio/2Dcur" ) == 0 ) {
			const char* cmd;
			args >> cmd;
			
			if( strcmp( cmd, "set" ) == 0 ){	
				int32 s_id;
				float x, y, X, Y, m;
					
				args >> s_id >> x >> y >> X >> Y >> m >> EndMessage;
				listener->updateTuioCur((int)s_id,x,y,X,Y,m);

			} else if( strcmp( cmd, "alive" ) == 0 ){

				int32 s_id;
				while(!args.Eos()) {
					args >> s_id;
					cursorBuffer.push_back((unsigned int)s_id);
					
					std::list<unsigned int>::iterator iter;
					iter = find(aliveCursorList.begin(), aliveCursorList.end(), (unsigned int)s_id); 
					if (iter != aliveCursorList.end()) aliveCursorList.erase(iter);
					else listener->addTuioCur((unsigned int)s_id);
				}
				args >> EndMessage;
				
				std::list<unsigned int>::iterator alive_iter;
				for (alive_iter=aliveCursorList.begin(); alive_iter != aliveCursorList.end(); alive_iter++) {
					listener->removeTuioCur(*alive_iter);
				}
			
				//std::list<unsigned int> newCursorList = aliveCursorList;
				aliveCursorList = cursorBuffer;
			
				// recycling the list
				//cursorBuffer = newCursorList;
				cursorBuffer.clear();
			}
		}
	} catch( Exception& e ){
		std::cout << "error while parsing message: "<< msg.AddressPattern() << ": " << e.what() << "\n";
	}
}

void TuioClient::ProcessPacket( const char *data, int size, const IpEndpointName& remoteEndpoint ) {
	if (listener==NULL) return;
	ReceivedPacket p( data, size );
	if(p.IsBundle()) ProcessBundle( ReceivedBundle(p), remoteEndpoint);
        else ProcessMessage( ReceivedMessage(p), remoteEndpoint);
}

void TuioClient::start(bool lk) {

	locked = lk;
	if (!locked) {
		#ifndef WIN32
		pthread_create(&thread , NULL, ThreadFunc, this);
		#else
		DWORD threadId;
		thread = CreateThread( 0, 0, ThreadFunc, this, 0, &threadId );
		#endif
	} else socket->Run();
}

void TuioClient::stop() {
	socket->Break();

	if (!locked) {
		#ifdef WIN32
		if( thread ) CloseHandle( thread );
		#endif
		thread = 0;
		locked = false;
	}
}

void TuioClient::addTuioListener(TuioListener *l) {
	listener = l;
}

void TuioClient::removeTuioListener() {
	listener = NULL;
}
